File: study/lua/src/app/scene/view/stage/PopupStarRankNode.lua
local ListViewCellBase = require("app.ui.ListViewCellBase")
local PopupStarRankNode = class("PopupStarRankNode", ListViewCellBase)
local Hero = require("app.config.hero")
local Color = require("app.utils.Color")
-- local Path = require("app.utils.Path")

PopupStarRankNode.RANK_BG_DARK = 4

function PopupStarRankNode:ctor(rankData)
    -- self._rankData = rankData

    self._textRank = nil    --名次
    self._textName = nil    --名字
    self._textRound = nil   --章节
    self._textStar = nil    --星数
    self._imageBG = nil     --背景图片
    -- self._imageRank = nil   --排行图标，前3
    self._rank = nil        --排行图标

	local resource = {
		file = Path.getCSB("PopupStarRankNode", "stage"),
		binding = {
            _panelBase = {
				events = {{event = "touch", method = "_onPanelClick"}}
			},
		}
	}
	PopupStarRankNode.super.ctor(self, resource)
end

function PopupStarRankNode:onCreate()
    self._size = self._panelBase:getContentSize()
    self:setContentSize(self._size)
    self._panelBase:setSwallowTouches(false)
    self._rank:setVisible(false)
    self._imageBG:setScale9Enabled(false)
end

function PopupStarRankNode:refreshInfo(rankData)
    self._rankData = rankData
    local rank = self._rankData:getRank()
    local officerLevel = self._rankData:getOfficer_level()
    self:_setNodeBG(rank)

    self._textRank:setString(rank)
    self._textStar:setString(self._rankData:getStar())
    self._textName:setString(self._rankData:getName())
    local chapter = self._rankData:getChapter()
    self._textRound:setString(Lang.get("mission_star_chapter", {num = chapter}))

    -- local commonColor = Color.getRankColor(rank)
    -- self._textRank:setColor(commonColor)
    -- self._textStar:setColor(commonColor)
    -- self._textRound:setColor(commonColor)

    self._textName:setColor(Color.getOfficialColor(officerLevel))
    require("yoka.utils.UIHelper").updateTextOfficialOutline(self._textName, officerLevel)
end

function PopupStarRankNode:_setNodeBG(rank)
    if rank < 4 then
        self._imageBG:setScale9Enabled(false)
        local pic = Path.getComplexRankUI("img_com_ranking0"..rank)
        self._imageBG:loadTexture(pic)

        self._imageBGLight:setVisible(true)

        self._imageBG:ignoreContentAdaptWithSize(true)
        self._textRank:setVisible(false)
        local icon = Path.getRankIcon(rank)
        self._rank:setRank(rank)
        self._rank:setVisible(true)
    else
        self._textRank:setVisible(true)
        -- self._imageRank:setVisible(false)
        self._rank:setVisible(false)

        self._imageBGLight:setVisible(false)
    end

    
    if rank >= 4 and rank % 2 == 1 then
        local pic = Path.getCommonRankUI("img_com_board_list01b")
        self._imageBG:loadTexture(pic)
        self._imageBG:ignoreContentAdaptWithSize(true)

        self._imageBG:setCapInsets(cc.rect(2, 3, 2, 1))
        self._imageBG:setScale9Enabled(true)
        self._imageBG:setContentSize(self._size)        
        -- self._imageBG:setVisible(true)
    elseif rank >= 4 and rank % 2 == 0 then
        local pic = Path.getCommonRankUI("img_com_board_list01a")
        self._imageBG:loadTexture(pic)
        self._imageBG:ignoreContentAdaptWithSize(true)
        self._imageBG:setCapInsets(cc.rect(2, 2, 1, 1))
        self._imageBG:setScale9Enabled(true)
        self._imageBG:setContentSize(self._size)        
        -- self._imageBG:setVisible(false)
    end
end

function PopupStarRankNode:_onPanelClick(sender)
	local offsetX = math.abs(sender:getTouchEndPosition().x - sender:getTouchBeganPosition().x)
	local offsetY = math.abs(sender:getTouchEndPosition().y - sender:getTouchBeganPosition().y)
	if offsetX < 20 and offsetY < 20  then
        local userId = self._rankData:getUser_id()
        if userId ~= G_UserData:getBase():getId() then
            G_UserData:getBase():c2sGetUserBaseInfo(userId)
        end
	end
end


return PopupStarRankNode

File: study/ts/script/scene/view/stage/PopupStarRankNode.ts
const { ccclass, property } = cc._decorator;

import CommonRankIcon from '../../../ui/component/CommonRankIcon'
import { handler } from '../../../utils/handler';
import { Lang } from '../../../lang/Lang';
import { Color } from '../../../utils/Color';
import UIHelper from '../../../utils/UIHelper';
import { Path } from '../../../utils/Path';
import { G_UserData } from '../../../init';

@ccclass
export default class PopupStarRankNode extends cc.Component {

    @property({
        type: cc.Node,
        visible: true
    })
    _panelBase: cc.Node = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageBG: cc.Sprite = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageBGLight: cc.Sprite = null;

    @property({
        type: CommonRankIcon,
        visible: true
    })
    _rank: CommonRankIcon = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textRank: cc.Label = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textName: cc.Label = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textRound: cc.Label = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textStar: cc.Label = null;

    private _size;
    private _rankData;
    onLoad() {
        this._size = this._panelBase.getContentSize();
        this.node.setContentSize(this._size);
        this._panelBase.on(cc.Node.EventType.TOUCH_END, handler(this, this._onPanelClick));
        //this._rank.node.active = (false);
    }
    refreshInfo(rankData) {
        this._rankData = rankData;
        var rank = this._rankData.getRank();
        var officerLevel = this._rankData.getOfficer_level();
        this._setNodeBG(rank);
        this._textRank.string = (rank);
        this._textStar.string = (this._rankData.getStar());
        this._textName.string = (this._rankData.getName());
        var chapter = this._rankData.getChapter();
        this._textRound.string = (Lang.get('mission_star_chapter', { num: chapter }));
        this._textName.node.color = (Color.getOfficialColor(officerLevel));
        UIHelper.updateTextOfficialOutline(this._textName.node, officerLevel);
    }
    _setNodeBG(rank) {
        if (rank < 4) {
            var pic = Path.getComplexRankUI('img_com_ranking0' + rank);
            UIHelper.loadTexture(this._imageBG, pic);
            this._imageBGLight.node.active = (true);
            // this._imageBG.sizeMode = cc.Sprite.SizeMode.RAW;
            this._textRank.node.active = (false);
            var icon = Path.getRankIcon(rank);
            this._rank.setRank(rank);
            this._rank.node.active =(true);
        } else {
            this._textRank.node.active =(true);
            this._rank.node.active =(false);
            this._imageBGLight.node.active =(false);
        }
        if (rank >= 4 && rank % 2 == 1) {
            var pic = Path.getCommonRankUI('img_com_board_list01b');
            UIHelper.loadTexture(this._imageBG, pic);
            // this._imageBG.sizeMode = cc.Sprite.SizeMode.RAW;
        } else if (rank >= 4 && rank % 2 == 0) {
            var pic = Path.getCommonRankUI('img_com_board_list01a');
            UIHelper.loadTexture(this._imageBG, pic);
            // this._imageBG.sizeMode = cc.Sprite.SizeMode.RAW;
        }
    }
    _onPanelClick(sender:cc.Event.EventTouch) {
        var offsetX = Math.abs(sender.getLocation().x - sender.getStartLocation().x);
        var offsetY = Math.abs(sender.getLocation().y - sender.getStartLocation().y);
        if (offsetX < 20 && offsetY < 20) {
            var userId = this._rankData.getUser_id();
            if (userId != G_UserData.getBase().getId()) {
                G_UserData.getBase().c2sGetUserBaseInfo(userId);
            }
        }
    }

}